blueocean-pipeline-editor/src/main/js/components/editor/EditorPipelineGraph.jsx

```diff
@@ -0,0 +1,603 @@
+// @flow
+
+import React, { Component, PropTypes } from 'react';
+import {getAddIconGroup} from './common';
+
+import type { StageInfo } from './common';
+
+// Dimensions used for layout, px
+export const defaultLayout = {
+    nodeSpacingH: 120,
+    nodeSpacingV: 70,
+    nodeRadius: 12,
+    curveRadius: 12,
+    connectorStrokeWidth: 3.5,
+    labelOffsetV: 25,
+    smallLabelOffsetV: 20
+};
+
+const nodeStrokeWidth = 3.5; //px
+
+// Typedefs
+
+type StageNodeInfo = {
+    // -- Shared with PlaceholderNodeInfo
+    key: string,
+    x: number,
+    y: number,
+    nodeId: number,
+    parentStage?: StageInfo,
+
+    // -- Marker
+    isPlaceholder: false,
+
+    // -- Unique
+    name: string,
+    stage: StageInfo
+};
+
+type PlaceholderNodeInfo = {
+    // -- Shared with StageNodeInfo
+    key: string,
+    x: number,
+    y: number,
+    nodeId: number,
+    parentStage?: StageInfo,
+
+    // -- Marker
+    isPlaceholder: true,
+
+    // -- Unique
+    type: "start" | "add"
+}
+
+// TODO: Attempt to extract a "common" node type with intersection operator to remove duplication
+
+type NodeInfo = StageNodeInfo | PlaceholderNodeInfo;
+
+type ConnectionInfo = [NodeInfo, NodeInfo];
+
+type LabelInfo = {
+    //key: string,
+    x: number,
+    y: number,
+    text: string,
+    node: NodeInfo,
+    stage?: StageInfo
+};
```

Comment:
Yes, I find it helps me more than it annoys me, which is basically as good as current type systems get